File: web/src/test/workoutRepository.test.ts
import { describe, expect, it } from 'vitest';

import { createFitLogDb } from '@/db/client';
import { applyMigrations } from '@/db/migrate';
import { migrations } from '@/db/migrations';
import { seedCatalog } from '@/data/catalogSeed';
import {
  addSet,
  deleteSet,
  finishSession,
  getActiveSession,
  getSessionDetail,
  importSessions,
  listSessions,
  startSession,
  updateSet,
  WorkoutError,
} from '@/data/workout';

import catalogSeed from '@shared/seed/catalog.json';

import { createNodeSqliteHarness } from './helpers/nodeSqlite';

type Db = ReturnType<typeof createFitLogDb>;

function first<T>(items: readonly T[]): T {
  const value = items[0];
  if (value === undefined) throw new Error('lista vacia');
  return value;
}

function createDatabase(): Db {
  const harness = createNodeSqliteHarness();
  applyMigrations(harness.runner, migrations);
  seedCatalog(harness.runner, catalogSeed, () => 1);
  return createFitLogDb(harness.client);
}

const exerciseA = first(catalogSeed.exercises.filter((item) => item.slug === 'press-banca-barra'));
const exerciseB = first(catalogSeed.exercises.filter((item) => item.slug === 'remo-barra'));

async function expectWorkoutError(promise: Promise<unknown>, code: string): Promise<void> {
  await expect(promise).rejects.toBeInstanceOf(WorkoutError);
  await promise.catch((error: unknown) => {
    expect((error as WorkoutError).code).toBe(code);
  });
}

describe('importación de entrenamientos', () => {
  const draft = {
    startedAtMs: 1_690_000_000_000,
    finishedAtMs: 1_690_000_600_000,
    notes: 'Huawei Health · Running · 5 km',
  };

  it('importa sesiones y saltea las que ya existen', async () => {
    const db = createDatabase();

    const firstRun = await importSessions(db, [draft]);
    expect(firstRun.imported).toBe(1);
    expect(firstRun.skipped).toBe(0);

    // Repetir la importación no duplica: la fecha de inicio ya está.
    const secondRun = await importSessions(db, [draft]);
    expect(secondRun.imported).toBe(0);
    expect(secondRun.skipped).toBe(1);

    const sessions = await listSessions(db);
    expect(sessions).toHaveLength(1);
    expect(first(sessions).notes).toBe('Huawei Health · Running · 5 km');
    expect(first(sessions).finishedAt).toBe(1_690_000_600_000);
  });

  it('importa solo lo nuevo de una exportación posterior', async () => {
    const db = createDatabase();
    await importSessions(db, [draft]);

    const nuevo = { startedAtMs: 1_700_000_000_000, finishedAtMs: null, notes: 'nuevo' };
    const result = await importSessions(db, [draft, nuevo]);

    expect(result.imported).toBe(1);
    expect(result.skipped).toBe(1);
    expect(await listSessions(db)).toHaveLength(2);
  });
});

describe('sesiones', () => {
  it('inicia y finaliza una sesion', async () => {
    const db = createDatabase();
    const session = await startSession(db, null, () => 1000);

    expect(session.startedAt).toBe(1000);
    expect(session.finishedAt).toBeNull();

    const active = await getActiveSession(db);
    expect(active?.id).toBe(session.id);

    await finishSession(db, session.id, () => 2000);
    expect(await getActiveSession(db)).toBeNull();

    const history = await listSessions(db);
    expect(history).toHaveLength(1);
    expect(history[0]?.finishedAt).toBe(2000);
  });

  it('rechaza una segunda sesion activa', async () => {
    const db = createDatabase();
    await startSession(db, null, () => 1000);
    await expectWorkoutError(startSession(db, null, () => 2000), 'session_already_active');
  });

  it('rechaza finalizar una sesion inexistente o ya finalizada', async () => {
    const db = createDatabase();
    const session = await startSession(db, null, () => 1000);
    await finishSession(db, session.id, () => 2000);

    await expectWorkoutError(finishSession(db, session.id, () => 3000), 'session_not_active');
    await expectWorkoutError(
      finishSession(db, '01ARYZ6S41TSV4RRFFQ69G5FAV', () => 3000),
      'session_not_found'
    );
  });
});

describe('series', () => {
  it('registra series con indice automatico por ejercicio', async () => {
    const db = createDatabase();
    const session = await startSession(db, null, () => 1000);

    const first = await addSet(
      db,
      {
        sessionId: session.id,
        exerciseId: exerciseA.id,
        weightKg: 100,
        reps: 8,
        rir: 2,
        notes: null,
        isWarmup: false,
      },
      () => 2000
    );
    const second = await addSet(
      db,
      {
        sessionId: session.id,
        exerciseId: exerciseA.id,
        weightKg: 100,
        reps: 6,
        rir: null,
        notes: 'buena',
        isWarmup: false,
      },
      () => 3000
    );
    const otherExercise = await addSet(
      db,
      {
        sessionId: session.id,
        exerciseId: exerciseB.id,
        weightKg: 60,
        reps: 10,
        rir: null,
        notes: null,
        isWarmup: false,
      },
      () => 4000
    );

    expect(first.setIndex).toBe(1);
    expect(second.setIndex).toBe(2);
    expect(otherExercise.setIndex).toBe(1);
  });

  it('valida peso, repeticiones y ejercicio', async () => {
    const db = createDatabase();
    const session = await startSession(db, null, () => 1000);

    await expectWorkoutError(
      addSet(db, {
        sessionId: session.id,
        exerciseId: exerciseA.id,
        weightKg: -5,
        reps: 8,
        rir: null,
        notes: null,
        isWarmup: false,
      }),
      'invalid_input'
    );

    await expectWorkoutError(
      addSet(db, {
        sessionId: session.id,
        exerciseId: exerciseA.id,
        weightKg: 100,
        reps: -1,
        rir: null,
        notes: null,
        isWarmup: false,
      }),
      'invalid_input'
    );

    await expectWorkoutError(
      addSet(db, {
        sessionId: session.id,
        exerciseId: '01ARYZ6S41TSV4RRFFQ69G5FAV',
        weightKg: 100,
        reps: 8,
        rir: null,
        notes: null,
        isWarmup: false,
      }),
      'exercise_not_found'
    );
  });

  it('rechaza series en una sesion finalizada', async () => {
    const db = createDatabase();
    const session = await startSession(db, null, () => 1000);
    await finishSession(db, session.id, () => 2000);

    await expectWorkoutError(
      addSet(db, {
        sessionId: session.id,
        exerciseId: exerciseA.id,
        weightKg: 100,
        reps: 8,
        rir: null,
        notes: null,
        isWarmup: false,
      }),
      'session_not_active'
    );
  });

  it('edita y elimina series con borrado logico', async () => {
    const db = createDatabase();
    const session = await startSession(db, null, () => 1000);
    const set = await addSet(
      db,
      {
        sessionId: session.id,
        exerciseId: exerciseA.id,
        weightKg: 100,
        reps: 8,
        rir: 2,
        notes: null,
        isWarmup: false,
      },
      () => 2000
    );

    await updateSet(db, set.id, { weightKg: 102.5, reps: 8, rir: 1, notes: 'subida' }, () => 3000);

    let detail = await getSessionDetail(db, session.id);
    expect(detail.sets[0]?.weightKg).toBe(102.5);
    expect(detail.sets[0]?.rir).toBe(1);

    await deleteSet(db, set.id, () => 4000);
    detail = await getSessionDetail(db, session.id);
    expect(detail.sets).toHaveLength(0);
    expect(detail.session.summary.totalVolumeKg).toBe(0);
  });

  it('el borrado de una serie no cambia los indices de las demas', async () => {
    const db = createDatabase();
    const session = await startSession(db, null, () => 1000);

    const created = [];
    for (const reps of [10, 8, 6]) {
      created.push(
        await addSet(
          db,
          {
            sessionId: session.id,
            exerciseId: exerciseA.id,
            weightKg: 80,
            reps,
            rir: null,
            notes: null,
            isWarmup: false,
          },
          () => 2000
        )
      );
    }

    await deleteSet(db, first(created.slice(1)).id, () => 3000);
    const detail = await getSessionDetail(db, session.id);

    expect(detail.sets.map((set) => set.setIndex)).toEqual([1, 3]);
  });
});

describe('historial', () => {
  it('lista sesiones con resumen y detalle agrupado por ejercicio', async () => {
    const db = createDatabase();
    const older = await startSession(db, null, () => 1000);
    await addSet(db, {
      sessionId: older.id,
      exerciseId: exerciseA.id,
      weightKg: 100,
      reps: 8,
      rir: null,
      notes: null,
      isWarmup: false,
    });
    await finishSession(db, older.id, () => 1000 + 45 * 60_000);

    const newer = await startSession(db, null, () => 5000);
    await addSet(db, {
      sessionId: newer.id,
      exerciseId: exerciseB.id,
      weightKg: 60,
      reps: 10,
      rir: null,
      notes: null,
      isWarmup: false,
    });
    await addSet(db, {
      sessionId: newer.id,
      exerciseId: exerciseB.id,
      weightKg: 40,
      reps: 10,
      rir: null,
      notes: null,
      isWarmup: true,
    });

    const history = await listSessions(db);
    expect(history.map((entry) => entry.id)).toEqual([newer.id, older.id]);
    expect(history[0]?.summary.workingSets).toBe(1);
    expect(history[0]?.summary.totalVolumeKg).toBe(600);
    expect(history[0]?.finishedAt).toBeNull();
    expect(history[1]?.summary.totalVolumeKg).toBe(800);

    const detail = await getSessionDetail(db, newer.id);
    expect(detail.sets).toHaveLength(2);
    expect(detail.sets[0]?.exerciseName).toBe('Remo con barra');
    expect(detail.sets[0]?.isWarmup).toBe(false);
    expect(detail.sets[1]?.isWarmup).toBe(true);
  });

  it('devuelve un detalle vacio para una sesion sin series', async () => {
    const db = createDatabase();
    const session = await startSession(db, null, () => 1000);
    await finishSession(db, session.id, () => 2000);

    const detail = await getSessionDetail(db, session.id);
    expect(detail.sets).toHaveLength(0);
    expect(detail.session.summary.totalVolumeKg).toBe(0);
  });
});

File: android/app/src/test/java/com/fitlog/app/WorkoutRepositoryTest.kt
package com.fitlog.app

import android.content.Context
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import com.fitlog.app.data.AddSetInput
import com.fitlog.app.data.CatalogRepository
import com.fitlog.app.data.FitLogDatabase
import com.fitlog.app.data.UpdateSetInput
import com.fitlog.app.data.WorkoutErrorCode
import com.fitlog.app.data.WorkoutException
import com.fitlog.app.data.WorkoutRepository
import kotlinx.coroutines.test.runTest
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class WorkoutRepositoryTest {

    private lateinit var database: FitLogDatabase
    private lateinit var repository: WorkoutRepository
    private lateinit var exerciseA: String
    private lateinit var exerciseB: String

    private val seedJson: String by lazy {
        val context = ApplicationProvider.getApplicationContext<Context>()
        context.assets.open("seed/catalog.json").bufferedReader().use { it.readText() }
    }

    @Before
    fun setUp() = runTest {
        database = Room.inMemoryDatabaseBuilder(
            ApplicationProvider.getApplicationContext(),
            FitLogDatabase::class.java,
        ).allowMainThreadQueries().build()

        val catalog = CatalogRepository(
            dao = database.catalogDao(),
            seedJsonProvider = { seedJson },
        )
        catalog.ensureSeeded()

        repository = WorkoutRepository(
            dao = database.workoutDao(),
            routinesDao = database.routinesDao(),
            now = { 1_700_000_000_000 },
        )
        val exercises = catalog.loadCatalog().exercises
        exerciseA = exercises.first { it.slug == "press-banca-barra" }.id
        exerciseB = exercises.first { it.slug == "remo-barra" }.id
    }

    @After
    fun tearDown() {
        database.close()
    }

    @Test
    fun `expone el descanso objetivo de la rutina de la sesion`() = runTest {
        val routines = com.fitlog.app.data.RoutinesRepository(
            dao = database.routinesDao(),
            now = { 1_700_000_000_000 },
        )
        val routine = routines.createRoutine("Dia de empuje", null)
        routines.addExercise(
            routine.id,
            com.fitlog.app.data.RoutineExerciseInput(
                exerciseId = exerciseA,
                targetSets = 4,
                targetReps = 8,
                targetWeightKg = null,
                restSeconds = 120,
                notes = null,
            ),
        )
        routines.addExercise(
            routine.id,
            com.fitlog.app.data.RoutineExerciseInput(
                exerciseId = exerciseB,
                targetSets = 3,
                targetReps = 10,
                targetWeightKg = null,
                restSeconds = null,
                notes = null,
            ),
        )

        val session = repository.startSession(routine.id)
        assertEquals(mapOf(exerciseA to 120), repository.restTargets(session.routineId))
        assertEquals(emptyMap<String, Int>(), repository.restTargets(null))
    }

    @Test
    fun `la serie guarda su marca de tiempo`() = runTest {
        val session = repository.startSession()
        repository.addSet(
            AddSetInput(
                sessionId = session.id,
                exerciseId = exerciseA,
                weightKg = 60.0,
                reps = 10,
                rir = 2,
                notes = null,
                isWarmup = false,
            ),
        )

        val sets = repository.sessionDetail(session.id).sets
        assertEquals(1, sets.size)
        assertEquals(1_700_000_000_000, sets.first().createdAtMs)
    }

    @Test
    fun `importa sesiones salteando las que ya existen`() = runTest {
        val draft = WorkoutRepository.ImportedSession(
            startedAtMs = 1_690_000_000_000,
            finishedAtMs = 1_690_000_600_000,
            notes = "Huawei Health · Running · 5 km",
        )

        val first = repository.importSessions(listOf(draft))
        assertEquals(1, first.imported)
        assertEquals(0, first.skipped)

        // Repetir la importacion no duplica: la fecha de inicio ya esta.
        val second = repository.importSessions(listOf(draft))
        assertEquals(0, second.imported)
        assertEquals(1, second.skipped)

        val sessions = repository.sessions()
        assertEquals(1, sessions.size)
        assertEquals("Huawei Health · Running · 5 km", sessions.first().notes)
        assertEquals(1_690_000_600_000, sessions.first().finishedAt)
    }

    @Test
    fun `importa solo lo nuevo de una exportacion posterior`() = runTest {
        val viejo = WorkoutRepository.ImportedSession(1_690_000_000_000, 1_690_000_600_000, "viejo")
        repository.importSessions(listOf(viejo))

        val nuevo = WorkoutRepository.ImportedSession(1_700_000_000_000, 1_700_000_600_000, "nuevo")
        val result = repository.importSessions(listOf(viejo, nuevo))

        assertEquals(1, result.imported)
        assertEquals(1, result.skipped)
        assertEquals(2, repository.sessions().size)
    }

    @Test
    fun `inicia y finaliza una sesion`() = runTest {
        val session = repository.startSession()
        assertEquals(session.startedAt, 1_700_000_000_000)
        assertNull(session.finishedAt)
        assertEquals(session.id, repository.activeSession()?.id)

        repository.finishSession(session.id)
        assertNull(repository.activeSession())

        val history = repository.sessions()
        assertEquals(1, history.size)
        assertEquals(1_700_000_000_000, history.first().finishedAt)
    }

    @Test
    fun `rechaza una segunda sesion activa`() = runTest {
        repository.startSession()
        val error = runCatching { repository.startSession() }.exceptionOrNull()
        assertEquals(WorkoutErrorCode.SESSION_ALREADY_ACTIVE, (error as WorkoutException).code)
    }

    @Test
    fun `registra series con indice automatico por ejercicio`() = runTest {
        val session = repository.startSession()

        val first = repository.addSet(set(session.id, exerciseA, weight = 100.0, reps = 8))
        val second = repository.addSet(set(session.id, exerciseA, weight = 100.0, reps = 6))
        val other = repository.addSet(set(session.id, exerciseB, weight = 60.0, reps = 10))

        assertEquals(1, first.setIndex)
        assertEquals(2, second.setIndex)
        assertEquals(1, other.setIndex)
    }

    @Test
    fun `valida peso repeticiones rir y ejercicio`() = runTest {
        val session = repository.startSession()

        assertEquals(
            WorkoutErrorCode.INVALID_INPUT,
            runCatching { repository.addSet(set(session.id, exerciseA, weight = -5.0, reps = 8)) }
                .exceptionOrNull()
                .let { (it as WorkoutException).code },
        )
        assertEquals(
            WorkoutErrorCode.INVALID_INPUT,
            runCatching { repository.addSet(set(session.id, exerciseA, weight = 100.0, reps = -1)) }
                .exceptionOrNull()
                .let { (it as WorkoutException).code },
        )
        assertEquals(
            WorkoutErrorCode.INVALID_INPUT,
            runCatching {
                repository.addSet(
                    set(session.id, exerciseA, weight = 100.0, reps = 8).copy(rir = 15)
                )
            }.exceptionOrNull().let { (it as WorkoutException).code },
        )
        assertEquals(
            WorkoutErrorCode.EXERCISE_NOT_FOUND,
            runCatching {
                repository.addSet(set(session.id, "01ARYZ6S41TSV4RRFFQ69G5FAV", weight = 100.0, reps = 8))
            }.exceptionOrNull().let { (it as WorkoutException).code },
        )
    }

    @Test
    fun `rechaza series en una sesion finalizada`() = runTest {
        val session = repository.startSession()
        repository.finishSession(session.id)

        val error = runCatching {
            repository.addSet(set(session.id, exerciseA, weight = 100.0, reps = 8))
        }.exceptionOrNull()

        assertEquals(WorkoutErrorCode.SESSION_NOT_ACTIVE, (error as WorkoutException).code)
    }

    @Test
    fun `edita y elimina series con borrado logico`() = runTest {
        val session = repository.startSession()
        val created = repository.addSet(set(session.id, exerciseA, weight = 100.0, reps = 8))

        repository.updateSet(created.id, UpdateSetInput(weightKg = 102.5, reps = 8, rir = 1, notes = "subida"))

        var detail = repository.sessionDetail(session.id)
        assertEquals(102.5, detail.sets.first().weightKg ?: 0.0, 1e-9)
        assertEquals(1, detail.sets.first().rir)

        repository.deleteSet(created.id)
        detail = repository.sessionDetail(session.id)
        assertTrue(detail.sets.isEmpty())
        assertEquals(0.0, detail.session.summary.totalVolumeKg, 1e-9)
    }

    @Test
    fun `el borrado no cambia los indices de las demas series`() = runTest {
        val session = repository.startSession()
        val indices = mutableListOf<Int>()
        repeat(3) { indices.add(repository.addSet(set(session.id, exerciseA, weight = 80.0, reps = 10)).setIndex) }

        val all = repository.sessionDetail(session.id).sets
        repository.deleteSet(all[1].id)

        val remaining = repository.sessionDetail(session.id).sets.map { it.setIndex }
        assertEquals(listOf(1, 3), remaining)
    }

    @Test
    fun `el historial incluye resumen por sesion`() = runTest {
        val older = repository.startSession()
        repository.addSet(set(older.id, exerciseA, weight = 100.0, reps = 8, isWarmup = false))
        repository.finishSession(older.id)

        val newer = repository.startSession()
        repository.addSet(set(newer.id, exerciseB, weight = 60.0, reps = 10, isWarmup = false))
        repository.addSet(set(newer.id, exerciseB, weight = 40.0, reps = 10, isWarmup = true))

        val history = repository.sessions()
        assertEquals(2, history.size)
        val latest = history.first()
        assertEquals(newer.id, latest.id)
        assertEquals(1, latest.summary.workingSets)
        assertEquals(600.0, latest.summary.totalVolumeKg, 1e-9)
        assertNull(latest.finishedAt)

        assertEquals(800.0, history[1].summary.totalVolumeKg, 1e-9)
    }

    private fun set(
        sessionId: String,
        exerciseId: String,
        weight: Double?,
        reps: Int?,
        rir: Int? = null,
        isWarmup: Boolean = false,
    ) = AddSetInput(
        sessionId = sessionId,
        exerciseId = exerciseId,
        weightKg = weight,
        reps = reps,
        rir = rir,
        notes = null,
        isWarmup = isWarmup,
    )
}
